### DISCUSSION TITLE
Tradeoffs of using Docker Machine for production? Alternatives?

### DISCUSSION THREAD

**Discussion by mjclemente** [BretFisher/ama#174] [Q&A]
In your [DockerCon 2018 session](https://dockercon2018.hubs.vidyard.com/watch/k3Cv676wmxAwYDxbvcgcgC) (and the [associated dogvscat repo](https://github.com/BretFisher/dogvscat)) you recommend using Docker Machine for the solo/small teams deploying Swarm to production (while acknowledging that it's not a perfect solution).

What are the specific tradeoffs / issues with doing this? Is it that only one local machine can manage the Swarm? Are the approaches to sharing Docker Machine control actually workable? What are practical alternatives?

Thanks for all the help you provide for new and smaller container teams!

**Comment by BretFisher** [comment]
Two main issues that limit the scalability of DM:

1. UI for dealing with more than 10-20 machines becomes slow. Try provisioning 30 vm's with it and then do a `docker-machine ls` and you'll see.

2. No RBAC for the docker API endpoint it opens. This is more of a limitation of Docker CE that Docker EE resolves. Anyone who has the certs that DM generates has full root to the hosts.

2. Sharing the keys it generates for remote docker API. Those keys are in `~/.docker` and there are more than a few tools out there for managing them and moving them around between systems, but nothing that's simple and still supported that I see.

**Comment by mjclemente** [comment]
Thanks! Makes sense. Really appreciate you taking the time to share insights and help out developers just getting started with this. 👍
